modules/common/src/main/scala/com/snowplowanalytics/snowplow/rdbloader/common/transformation/WideField.scala

```diff
@@ -0,0 +1,110 @@
+
+import cats.implicits._
+import cats.Monad
+import cats.data.{EitherT, NonEmptyList}
+import cats.effect.Clock
+
+import com.snowplowanalytics.iglu.client.resolver.registries.RegistryLookup
+import com.snowplowanalytics.iglu.client.{ClientError, Resolver}
+import com.snowplowanalytics.iglu.core.{SchemaKey, SelfDescribingData}
+import com.snowplowanalytics.iglu.schemaddl.parquet.{CastError, Field, FieldValue}
+import com.snowplowanalytics.iglu.schemaddl.jsonschema.Schema
+import com.snowplowanalytics.iglu.schemaddl.jsonschema.circe.implicits._
+import com.snowplowanalytics.snowplow.analytics.scalasdk.Event
+import com.snowplowanalytics.snowplow.badrows._
+import com.snowplowanalytics.snowplow.rdbloader.common.LoaderMessage
+
+object WideField {
+
+  def forEntity[F[_]: Clock: Monad: RegistryLookup](resolver: Resolver[F],
+                                                    stype: LoaderMessage.TypesInfo.WideRow.Type,
+                                                    event: Event,
+                                                    processor: Processor): EitherT[F, BadRow, FieldValue] =
+    stype.snowplowEntity match {
+      case LoaderMessage.SnowplowEntity.SelfDescribingEvent =>
+        forUnstruct(resolver, stype, event, processor)
+      case LoaderMessage.SnowplowEntity.Context =>
+        forContexts(resolver, stype, event, processor)
+    }
+
+  def forUnstruct[F[_]: Clock: Monad: RegistryLookup](resolver: Resolver[F],
+                                                      stype: LoaderMessage.TypesInfo.WideRow.Type,
+                                                      event: Event,
+                                                      processor: Processor): EitherT[F, BadRow, FieldValue] = {
+    event.unstruct_event.data match {
+      case Some(SelfDescribingData(schemaKey, data)) if keysMatch(schemaKey, stype.schemaKey) =>
+        for {
+          schema <- getSchema(resolver, stype.schemaKey).leftMap(igluBadRow(event, processor, _))
+          field = Field.build("NOT_NEEDED", schema, false)
+          v <- EitherT.fromEither[F](FieldValue.cast(field)(data).toEither).leftMap(castingBadRow(event, processor, schemaKey))
+        } yield v
+      case _ => EitherT.rightT(FieldValue.NullValue)
+    }
+  }
+
+  def forContexts[F[_]: Clock: Monad: RegistryLookup](resolver: Resolver[F],
+                                                      stype: LoaderMessage.TypesInfo.WideRow.Type,
+                                                      event: Event,
+                                                      processor: Processor): EitherT[F, BadRow, FieldValue] = {
+    val rows = (event.contexts.data ::: event.derived_contexts.data)
+      .filter(sdd => keysMatch(sdd.schema, stype.schemaKey))
+    if (rows.nonEmpty) {
+      for {
+        schema <- getSchema(resolver, stype.schemaKey).leftMap(igluBadRow(event, processor, _))
+        field = Field.build("NOT_NEEDED", schema, true)
```

Comment:
I'm thinking about those `Field.build` here... we do exact same thing in `SparkSchema`, right? Maybe we could build all fields beforehand (we have all necessary types?) and then pass them to cast values and build schema?

Reply:
I agree this bit is still very ugly --  I missed this bit in my recent tidy up.  @pondzix would you like to take a go at making it neater?

Reply:
I created [PR](https://github.com/snowplow/snowplow-rdb-loader/pull/871) with proposed changes